src/main/java/alex/mojaki/s3upload/StreamTransferManager.java

```diff
@@ -343,14 +366,39 @@ public void complete() {
                         uploadId,
                         partETags);
                 customiseCompleteRequest(completeRequest);
-                s3Client.completeMultipartUpload(completeRequest);
+                CompleteMultipartUploadResult completeMultipartUploadResult = s3Client.completeMultipartUpload(completeRequest);
+                if (checkIntegrity) {
+                    checkCompleteFileIntegrity(completeMultipartUploadResult.getETag(), partETags);
+                }
             }
             log.info("{}: Completed", this);
         } catch (Throwable e) {
             throw abort(e);
         }
     }
 
+    private void checkCompleteFileIntegrity(String s3ObjectETag, List<PartETag> partsETags) throws NoSuchAlgorithmException {
+        List<PartETag> parts = new ArrayList<PartETag>(partsETags);
+        Collections.sort(parts, new PartNumberComparator());
+        String expectedETag = computeCompleteFileETag(parts);
+        if (!expectedETag.equals(s3ObjectETag)) {
+            throw new RuntimeException(String.format("Integrity check failed. Expected ETag: %s but actual is %s", expectedETag, s3ObjectETag));
+        }
+    }
+
+    private String computeCompleteFileETag(List<PartETag> parts) throws NoSuchAlgorithmException {
+        // When S3 combines the parts of a multipart upload into the final object, the ETag value is set to the
+        // hex-encoded MD5 hash of the concatenated binary-encoded (raw bytes) MD5 hashes of each part followed by
+        // "-" and the number of parts.
+        MessageDigest md = MessageDigest.getInstance("MD5");
+        md.reset();
+        for (PartETag partETag : parts) {
+            md.update(BinaryUtils.fromHex(partETag.getETag()));
+        }
+        // Represent byte array as a 32-digit number hexadecimal format followed by "-<partCount>".
```

Comment:
Can you link to a source for these comments?

Reply:
I wish I could add a link to the official documentation. I can only link some more or less reliable sources such as:
 * https://stackoverflow.com/questions/42208998/aws-s3-upload-integrity
 * https://www.savjee.be/2015/10/Verifying-Amazon-S3-multi-part-uploads-with-ETag-hash/
 * https://teppen.io/2018/06/23/aws_s3_etags/
 * https://github.com/lambfrier/calc_s3_etag
 * https://stackoverflow.com/a/19304527/794158 and https://github.com/antespi/s3md5